pkg/applicationserver/applicationserver.go

```diff
@@ -512,35 +513,46 @@ func (d downlinkQueueOperation) shouldSkip(sessionKeyID []byte) bool {
 
 const maxDownlinkQueueOperationAttempts = 50
 
-func (as *ApplicationServer) attemptDownlinkQueueOp(ctx context.Context, dev *ttnpb.EndDevice, link *ttnpb.ApplicationLink, peer cluster.Peer, op downlinkQueueOperation) ([]string, error) {
+func (as *ApplicationServer) attemptDownlinkQueueOp(ctx context.Context, dev *ttnpb.EndDevice, link *ttnpb.ApplicationLink, peer cluster.Peer, op downlinkQueueOperation) (mask []string, err error) {
+	resultCallbackItems := op.Items
+	if op.ResultFunc != nil {
+		defer func() {
+			op.ResultFunc(resultCallbackItems, err)
+		}()
+	}
+
+	sessions := make([]*ttnpb.Session, 0, 2)
```

Comment:
This was in the loop because the sessions may be updated by `as.buildSessionsFromError`

Also you cannot encrypt ahead of time, since the frame counter may have advanced.